File: 2021/day-05.py
SEGMENTS = []
with open('input/day-05-input.txt') as f:
    for line in f:
        SEGMENTS.append([[int(x) for x in y.split(',')]
                         for y in line.strip().split(' -> ')])


def add_point(points, x, y):
    p = ','.join((str(x), str(y)))
    if p in points:
        points[p] += 1
    else:
        points[p] = 1


def part1():
    points = {}

    for a, b in SEGMENTS:
        if a[0] == b[0]:
            p = min(a[1], b[1])
            q = max(a[1], b[1]) + 1
            for i in range(p, q):
                add_point(points, a[0], i)

        elif a[1] == b[1]:
            p = min(a[0], b[0])
            q = max(a[0], b[0]) + 1
            for i in range(p, q):
                add_point(points, i, a[1])

    return sum(x > 1 for x in list(points.values())), points

#---- Part 2 -------------------------------

def is_diagonal(a, b):
    return abs(a[0] - b[0]) == abs(a[1] - b[1])


def get_diagonal_points(a, b):
    x1, y1, x2, y2 = a[0], a[1], b[0], b[1]
    if x1 > x2:
        x1, y1, x2, y2 = x2, y2, x1, y1
    slope = (y2 - y1) // (x2 - x1)
    result = []

    for i, j in zip(range(x1, x2), range(y1, y2, slope)):
        result.append([i, j])
    result.append([x2, y2])  # add end point

    return result


def part2():
    points = part1()[1]
    for a, b in SEGMENTS:
        if is_diagonal(a, b):
            for p in get_diagonal_points(a, b):
                add_point(points, p[0], p[1])
    return sum(x > 1 for x in list(points.values()))


print(part1()[0]) # 4421
print(part2()) # 18674

File: 2025/pascal/day-05.pas
Program day_05;

{$mode objfpc}{$H+}{$J-}{$R+}{$Q+}

Uses SysUtils, StrUtils, Math;

const
   INPUT = '../input/day-05-input.txt';

type
   TRange = Record
      a, b: Int64;
   end;
   TFreshArray = Array of TRange;
   TAvailableArray = Array of Int64;


procedure GetInput(Var fresh: TFreshArray; Var available: TAvailableArray);
var
   tfIn: textFile;
   line: String;
   limits: Array of String;
   range: TRange;
begin
   AssignFile(tfIn, INPUT);
   SetLength(fresh, 0);
   SetLength(available, 0);

   try
      reset(tfIn);
      while not eof(tfIn) do
         begin
            ReadLn(tfIn, line);
            if line.Contains('-') then
               begin
                  limits := line.Split('-');
                  range.a := StrToInt64(limits[0]);
                  range.b := StrToInt64(limits[1]);
                  SetLength(fresh, Length(fresh) + 1);
                  fresh[High(fresh)] := range;
               end
            else
               if Length(line) > 0 then
                  begin
                     SetLength(available, Length(available) + 1);
                     available[High(available)] := StrToInt64(line);
                  end;
         end;
      CloseFile(tfIn);

   except
      on E: EInOutError do WriteLn('IOerror: ', E.Message);
   end;
end;


procedure Part_1(Const fresh: TFreshArray; Const available: TAvailableArray);
var
   x: Int64;
   r: TRange;
   count: Integer;
begin
   count := 0;
   for x in available do
      for r in fresh do
         if (r.a <= x) and (x <= r.b) then
            begin
               Inc(count);
               break;
            end;
   WriteLn(count);
end;


procedure SortIntervals(Var fresh: TFreshArray);
var
   i, j: Integer;
   temp: TRange;
begin
   for i := Low(fresh) to High(fresh) - 1 do
      for j := i + 1 to High(fresh) do
         if fresh[i].a > fresh[j].a then
            begin
               temp := fresh[i];
               fresh[i] := fresh[j];
               fresh[j] := temp;
            end;
end;


function MergeIntervals(Var fresh: TFreshArray): TFreshArray;
var
   merged: TFreshArray;
   r_high: Int64;
   r: TRange;
begin
   SetLength(merged, 0);
   SortIntervals(fresh);
   for r in fresh do
      if Length(merged) = 0 then
         begin
            SetLength(merged, Length(merged) + 1);
            merged[High(merged)] := r;
         end
      else
         begin
            r_high := merged[High(merged)].b;
            if r.a > (1 + r_high) then
               begin
                  SetLength(merged, Length(merged) + 1);
                  merged[High(merged)] := r;
               end
            else
               merged[High(merged)].b := Max(r.b, r_high);
         end;
   result := merged;
end;


procedure Part_2(fresh: TFreshArray);
var
   sum: Int64;
   r: TRange;
begin
   sum := 0;
   for r in MergeIntervals(fresh) do
      Inc(sum, 1 + r.b - r.a);
   WriteLn(sum);
end;


// main
var
   fresh: TFreshArray;
   available: TAvailableArray;

begin
   GetInput(fresh, available);
   Part_1(fresh, available);
   Part_2(fresh);
end.
